apps/verticalhome/js/pin_app_navigation.js

```diff
@@ -0,0 +1,294 @@
+
+        }
+
+      };
+
+      if (this.preventDef) {
+        e.preventDefault();
+      }
+      
+      function preVerticalNavigation () {
+        if (this.middleElem) {
+          document.querySelector('.middle').classList.remove('middle');
+          this.middleElem = false;
+        }
+
+
+        var toRemoved = this.elemList.querySelector('.removed');
+        var toRemoveOutFocus = this.elemList.querySelector('.out-focus');
+
+        if (toRemoved) {
+          this.elemList.removeChild(toRemoved);
+        }
+
+        if (toRemoveOutFocus) {
+          toRemoveOutFocus.classList.remove('out-focus');
+        }
+      };
+
+      if (this.preventDef){
+        e.preventDefault();
+      }
+
+
+      switch(e.key) {
+        case 'ArrowUp':
+
+          preVerticalNavigation.call(this);
+          document.getElementById('clock').style.opacity = 0;
+
+          if (this.elemList.dataset.scrollup) {
+            this.elemList.style.top = this.elemList.dataset.scrollup;
+          } else {
+            this.elemList.style.top = '-6rem';
+            this.elemList.dataset.scrollup = this.elemList.style.top;
+          }
+
+          var deltaElemsToStartScroll = 3;
+
+          this.elements[this.selectedElemIndex].classList.remove('selected');
+          if (this.selectedElemIndex >= 0 && this.selectedElemIndex < 4) {
+
+            cycleElements.call(this, (deltaElemsToStartScroll - this.selectedElemIndex));
+            this.selectedElemIndex += deltaElemsToStartScroll - this.selectedElemIndex;
+
+            this.selectedElemIndex--;
+            this.elements[this.selectedElemIndex].classList.add('selected');
+
+          } else {
+
+            this.selectedElemIndex--;
+            this.elements[this.selectedElemIndex].classList.add('selected');
+
+            if (this.selectedElemIndex > 3) {
+              this.elemList.style.top = parseFloat(this.elemList.style.top) + 6 + 'rem';
+              this.elemList.dataset.scrollup = this.elemList.style.top;
+            }
+
+          }
+
+          var lastElem = this.elements[this.elements.length - 1];
+          var clonedElem = lastElem.cloneNode(true);
+
+
+          this.elemList.insertBefore(clonedElem, this.elements[0]);
+          this.elements.splice(0, 0, clonedElem);
+          this.elements[0].classList.add('out-focus');
+          this.selectedElemIndex++;
+
+          lastElem.classList.add('removed');
+          this.elements.splice(-1, 1);
+
+          this.prevVerticalKey = 'up';
+
+
+          break;
+
+        case 'ArrowDown':
+
+          preVerticalNavigation.call(this);
+          document.getElementById('clock').style.opacity = 0;
+
+          this.elements[this.selectedElemIndex].classList.remove('selected');
+          if (this.selectedElemIndex == (this.elements.length - 3)) {
+            if (!this.elemList.dataset.scrolldown) {
+              this.elemList.style.top = parseFloat(this.elemList.style.top) - 6 + 'rem';
```

Comment:
I don't know why you're doing this, but I wonder if a CSS transform:translate() would be better than messing with style.top like this.